vortex-array/src/array/bool/compute/take.rs

```diff
@@ -1,55 +1,91 @@
 use arrow_buffer::BooleanBuffer;
 use itertools::Itertools;
 use num_traits::AsPrimitive;
-use vortex_dtype::match_each_integer_ptype;
+use vortex_dtype::{match_each_integer_ptype, NativePType};
 use vortex_error::VortexResult;
+use vortex_mask::Mask;
+use vortex_scalar::Scalar;
 
-use crate::array::{BoolArray, BoolEncoding};
-use crate::compute::TakeFn;
+use crate::array::{BoolArray, BoolEncoding, ConstantArray, PrimitiveArray};
+use crate::compute::{fill_null, TakeFn};
 use crate::variants::PrimitiveArrayTrait;
 use crate::{Array, IntoArray, IntoArrayVariant};
 
 impl TakeFn<BoolArray> for BoolEncoding {
     fn take(&self, array: &BoolArray, indices: &Array) -> VortexResult<Array> {
-        let validity = array.validity();
-        let indices = indices.clone().into_primitive()?;
-
-        // For boolean arrays that roughly fit into a single page (at least, on Linux), it's worth
-        // the overhead to convert to a Vec<bool>.
-        let buffer = if array.len() <= 4096 {
-            let bools = array.boolean_buffer().into_iter().collect_vec();
-            match_each_integer_ptype!(indices.ptype(), |$I| {
-                take_byte_bool(bools, indices.as_slice::<$I>())
-            })
-        } else {
-            match_each_integer_ptype!(indices.ptype(), |$I| {
-                take_bool(&array.boolean_buffer(), indices.as_slice::<$I>())
-            })
+        let indices_nulls_zeroed = match indices.validity_mask()? {
+            Mask::AllTrue(_) => indices.clone(),
+            Mask::AllFalse(_) => {
+                return Ok(
+                    ConstantArray::new(Scalar::null_typed::<bool>(), indices.len()).into_array(),
+                )
+            }
+            Mask::Values(_) => fill_null(indices, Scalar::from(0).cast(indices.dtype())?)?,
```

Comment:
I don't think this logic is really bool-specific... should it live in the top-level `take` impl? 

Also, there's a super niche edge case that should definitely be handled in the top-level where `array` is empty, some of the indices are not null, and the first index is null => we'll (hopefully) get OOB for index 0 even though they didn't ask for 0.

Reply:
you can't apply validity at top level so you have to dispatch this, you could copy validity, fill_null, take and then mask

Reply:
Agreed. I like your suggestion, but I cannot easily do it in this PR. In particular:

If we lift this work to the top-level, then I must pass a non-nullable indices array to `TakeFn::take`. Each encoding is now unable to use the indices' (in)validity to construct the result arrays. Moreover, we don't yet have `mask` (🤦), so the top-level `take` cannot apply the validity.

I can follow up the mask PR with an attempt to lift validity manipulation to the top-level. I'm not certain that's the right choice for every array. A second option is to change `TakeFn::take` to accept three arguments: array, (non-nullable) indices, and validity.

Reply:
Basically what Rob said 😉, gotta refresh my GitHub pages more often.